text/0000-process-type-arguments.md

```diff
@@ -0,0 +1,221 @@
+For example, the following does not print `hello world`.
+```
+[[process]]
+command = "echo"
+args = ["hello", "world"]
+```
+Instead, users must construct a string command that includes arguments
+```
+[[process]]
+command = "echo hello world"
+args = []
+```
+This already causes confusion for users (https://github.com/buildpacks/lifecycle/issues/281). The problem is compounded if we allow users to supply additional arguments. We cannot reliably append arguments to the `command` string without introducing tokenization or shell parsing errors. Constructing a `command` that accepts args in the way they are currently provided to `bash` is confusing and difficult (buildpack authors likely will not do this correctly).
+
+This RFC proposes changes to the `launcher`'s handling of args to make the execution of the following process
+```
+[[process]]
+command = "<CMD>"
+args = ["ARG1", "ARG2"]
+```
+produce results equivalent to the execution of `"<CMD>" "<ARG1>" "<ARG2>"` in a shell
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Terminology
+* shell process - a process with `direct=false`
+
+## Philosophy
+The goal is to make the runtime usage of a CNB container as similar as possible to the runtime usage on a non-CNB container, without removing functionality or violating assumptions that have been firmly established in the buildpack community (e.g. defaulting to `web`).
+
+Because the `launcher` must occupy the container `entrypoint`, we will make the `launcher` a multicall binary that derives the intended process-type, or lack there of, from `argv0`. This allows for intuitive use of both `entrypoint` and `cmd`: `entrypoint` will be used to select a process and `cmd` will be used to supply additional argument to that process.
+
+## Behavior
+
+### Exporter
+When `exporter` creates an app image it will create one symlink per buildpack-contributed process type. The symlink files will be named `/cnb/process/<process-type>` and will point at the launcher.
+
+The `exporter` will prepend `/cnb/process` to the image `PATH`.
+The `exporter` will set the app image `entrypoint` to the symlink matching the default process.
+
+Currently, it is possible to set the default process-type to a non-existent process-type. This will no longer be allowed. If the default process does not exist, the `exporter` will warn and set the entrypoint to `/cnb/lifecycle/launcher`.
+
+### Selecting a process type at runtime
+Users will no longer be able to pass a process-type to the launcher as a positional argument or with the `CNB_PROCESS_TYPE` environment variable. Instead, they must set the `entrypoint` in the running container to select a process type other than the default.
+
+For Example:
+`docker run --entrypoint web <image>`
+`docker run --entrypoint /cnb/process/web <image>`
+Will both launch the process with `type` `web`
+
+Buildpack-contributed processes will be forbidden from having `type` equal to `"launcher"`.
```

Comment:
Are we saying that buildpack authors will have no way of getting the `--entrypoint` to be `launcher` if any other buildpack defined a process?

Reply:
We don't want to allow ambiguity in `$0` between a process with type `launcher` and the normal `launcher`